core/application/impl/app_config_impl.cpp

```diff
@@ -0,0 +1,332 @@
+    Handler handler;
+  };
+
+  template <typename T, typename Func>
+  inline void find_argument(boost::program_options::variables_map &vm,
+                            char const *name,
+                            Func &&f) {
+    assert(nullptr != name);
+    auto it = vm.find(name);
+    if (it != vm.end()) std::forward<Func>(f)(it->second.as<T>());
+  }
+
+  const std::string def_rpc_http_host = "0.0.0.0";
+  const std::string def_rpc_ws_host = "0.0.0.0";
+  const uint16_t def_rpc_http_port = 40363;
+  const uint16_t def_rpc_ws_port = 40364;
+  const uint16_t def_p2p_port = 30363;
+  const int def_verbosity = 2;
+  const bool def_is_only_finalizing = false;
+}  // namespace
+
+namespace kagome::application {
+
+  void AppConfigurationImpl::file_deleter(FILE *f) {
+    std::fclose(f);
+  }
+
+  AppConfigurationImpl::AppConfigurationImpl(kagome::common::Logger logger)
+      : logger_(std::move(logger)),
+        rpc_http_host_(def_rpc_http_host),
+        rpc_ws_host_(def_rpc_ws_host),
+        rpc_http_port_(def_rpc_http_port),
+        rpc_ws_port_(def_rpc_ws_port),
+        p2p_port_(def_p2p_port),
+        verbosity_(static_cast<spdlog::level::level_enum>(def_verbosity)),
+        is_only_finalizing_(def_is_only_finalizing) {}
+
+  AppConfigurationImpl::FilePtr AppConfigurationImpl::open_file(
+      const std::string &filepath) {
+    assert(!filepath.empty());
+    return AppConfigurationImpl::FilePtr(std::fopen(filepath.c_str(), "r"),
+                                         &file_deleter);
+  }
+
+  [[maybe_unused]] size_t AppConfigurationImpl::get_file_size(
+      const std::string &filepath) {
+    struct stat statbuf;
+    if (-1 == stat(filepath.c_str(), &statbuf)) {
+      logger_->error("Read configuration file metadata failed: {}", filepath);
+      exit(EXIT_FAILURE);
+    }
+    return static_cast<size_t>(statbuf.st_size);
+  }
+
+  bool AppConfigurationImpl::load_str(const rapidjson::Value &val,
+                                      char const *name,
+                                      std::string &target) {
+    auto m = val.FindMember(name);
+    if (val.MemberEnd() != m && m->value.IsString()) {
+      target.assign(m->value.GetString(), m->value.GetStringLength());
+      return true;
+    }
+    return false;
+  }
+
+  bool AppConfigurationImpl::load_bool(const rapidjson::Value &val,
+                                       char const *name,
+                                       bool &target) {
+    auto m = val.FindMember(name);
+    if (val.MemberEnd() != m && m->value.IsBool()) {
+      target = m->value.GetBool();
+      return true;
+    }
+    return false;
+  }
+
+  bool AppConfigurationImpl::load_u16(const rapidjson::Value &val,
+                                      char const *name,
+                                      uint16_t &target) {
+    auto m = val.FindMember(name);
+    if (val.MemberEnd() != m && m->value.IsInt()) {
+      target = static_cast<uint16_t>(m->value.GetInt());
```

Comment:
It can be insecure, because is narrowing cast. Need to check value, i.e. by `std::numeric_limits`